File: samples/SampleCounterV4/Components/CounterApp.cs
using SampleCounterV4.Inferno;
using SampleCounterV4.Models;
using static SampleCounterV4.Mvu.Ui;

namespace SampleCounterV4.Components;

/// <summary>
/// Stateful Inferno component. Subclasses Inferno's <c>Component</c>
/// (imported from <c>inferno</c>) and overrides <see cref="Render"/>
/// to project the current state into a virtual-DOM tree built by the
/// JSX-flavored widget facade.
/// </summary>
public sealed class CounterApp : Component<EmptyProps, Counter>
{
    public override InfernoElement Render()
    {
        var state = State ?? Counter.Zero;

        return Column(
            gap: 12,
            Heading($"Count: {state.Count}", level: 1),
            Row(
                gap: 8,
                Button("➖", onClick: () => SetState(state.Decrement())),
                Button("➕", onClick: () => SetState(state.Increment())),
                Button("Reset", onClick: () => SetState(Counter.Zero))
            )
        );
    }
}

File: targets/js/sample-counter-v5/lib/components/counter-app.js
// Override of the generated CounterApp module. Two issues block the
// generated output today:
//   - Inferno's `Component<P, S>` declares `state: S | null`; Metano
//     drops the C# null-forgiving operator (`State!.Count`) so the
//     emitted body fails strict-null checks.
//   - `[Emit]` template type-arg placeholders (#189) aren't shipped
//     yet; once they are, this file goes away.
import { Component } from "inferno";
import { createElement } from "inferno-create-element";
import { Counter } from "#/models";
import { Button, Column, Heading, Row } from "#/mvu";
export class CounterApp extends Component {
    constructor() {
        super();
        this.state = Counter.zero;
    }
    render() {
        const state = this.state;
        return Column({
            gap: 12,
            children: [
                Heading({ content: `Count: ${state.count}` }),
                Row({
                    gap: 8,
                    children: [
                        Button({
                            label: "➖",
                            onClick: () => this.setState(state.decrement()),
                        }),
                        Button({
                            label: "➕",
                            onClick: () => this.setState(state.increment()),
                        }),
                        Button({
                            label: "Reset",
                            onClick: () => this.setState(Counter.zero),
                        }),
                    ],
                }),
            ],
        });
    }
}
